src/Core/Geometry/IndexedGeometry.cpp

```diff
@@ -200,22 +200,31 @@ void MultiIndexedGeometry::unlockFirstLayerOccurrence( const LayerSemanticCollec
 void MultiIndexedGeometry::unlockLayer( const LayerKeyType& layerKey ) {
     auto& p = m_indices.at( layerKey );
     CORE_ASSERT( p.first, "try to release unlocked layer" );
-    p.first = true;
+    p.first = false;
     notify();
 }
 
 //////////////////////////////////////////////////////////////////////
 //////////////////////////////////////////////////////////////////////
 
-bool MultiIndexedGeometry::addLayer( std::unique_ptr<GeometryIndexLayerBase>&& layer,
-                                     const std::string& layerName ) {
+std::pair<bool, GeometryIndexLayerBase&>
+MultiIndexedGeometry::addLayer( std::unique_ptr<GeometryIndexLayerBase>&& layer,
+                                const bool withLock,
+                                const std::string& layerName ) {
     LayerKeyType key { layer->semantics(), layerName };
-    if ( m_indices.find( key ) != m_indices.end() ) return false;
+    std::pair<LayerKeyType, EntryType> elt { key, std::make_pair( false, layer.get() ) };
+    auto [pos, inserted] = m_indices.insert( std::move( elt ) );
+    if ( inserted ) { layer.release(); }
+    notify();
 
-    m_indices.insert( { key, std::make_pair( false, layer.release() ) } );
+    if ( withLock ) {
+        CORE_ASSERT( !pos->second.first, "try to get already locked layer" );
+        pos->second.first = true;
+    }
+    /// TODO What happens to the unique ptr when it is not inserted ? (The problem was the same in
+    /// the previous version).
```

Comment:
it's deleted, isn't it ?

Reply:
Yes it is. 
Here, "previous version" refer to the actual implementation in release-candidate.
I'll change the comment to state that the caller must ensure the pointer could be safely deleted.